### PR TITLE
Bool tensor creation (cpu)

### PR DESCRIPTION
**Pull request by izdeby**
This PR enables bool tensor creation and some basic operations for the CPU backend. This is a part of Bool Tensor feature implementation work. The whole plan looks like this: 
    1. Storage Implementation [Done] 
    2. Tensor Creation.
        a) CPU (this PR)
        b) CUDA
    3. Tensor Conversions.
    4. Tensor Indexing.
    5. Tensor Operations.
    6. Back compatibility related changes.

**Change**:
Enable CPU tensors and these operations: 
- torch.zeros
- torch.tensor
- torch.ones
- torch.randint
- torch.full
- torch.full_like
- torch.empty
- torch.empty_like

**Tested via**:
1) unit tests 

2)
torch.zeros(2,2, dtype=torch.bool)
torch.tensor([True, False], dtype=torch.bool)
torch.tensor([-1, -1.1, 0, 1, 1.1, 2], dtype=torch.bool)
torch.ones([1,2], dtype=torch.bool)
torch.randint(10, (2, 2), dtype=torch.bool)
torch.full((2, 3), True, dtype=torch.bool)
torch.empty(4, dtype=torch.bool)

a = torch.tensor([0,0,1])
b = torch.full_like(a, True)

### CODE REVIEW

File: torch/csrc/utils/tensor_types.cpp [hunk 1]
```diff
@@ -72,7 +72,7 @@ std::vector<std::pair<Backend, ScalarType>> all_declared_types() {
   // can't easily iterate over enum classes
   std::vector<Backend> backends = { Backend::CPU, Backend::CUDA, Backend::SparseCPU, Backend::SparseCUDA };
   std::vector<ScalarType> scalar_types = { ScalarType::Byte, ScalarType::Char, ScalarType::Double, ScalarType::Float,
-                                           ScalarType::Int, ScalarType::Long, ScalarType::Short, ScalarType::Half};
+                                           ScalarType::Int, ScalarType::Long, ScalarType::Short, ScalarType::Half, ScalarType::Bool};
```
Comment (gchanan) [inline torch/csrc/utils/tensor_types.cpp]: are we generating a sparse half type?  IMO for now we should keep this the same as SparseHalf, i.e. not supported.

File: torch/csrc/utils/python_numbers.h [hunk 2]
```diff
@@ -59,6 +63,18 @@ inline int64_t THPUtils_unpackLong(PyObject* obj) {
   return (int64_t)value;
 }
 
+inline int64_t THPUtils_unpackBool(PyObject* obj) {
+  int overflow;
```
Comment (gchanan) [inline torch/csrc/utils/python_numbers.h]: where is this used?  I.e. why did you use this instead of just checking if you have Py_TRUE or Py_False?  This is always called after checkBool, right so you only have those values?

Also I'm not sure if this will work, because in python 2 bools are an instance of int, not long.

File: test/common_utils.py [hunk 3]
```diff
@@ -388,6 +388,11 @@ def assertEqual(self, x, y, prec=None, message='', allow_inf=False):
             self.assertEqual(x.item(), y, prec, message, allow_inf)
         elif isinstance(y, torch.Tensor) and isinstance(x, Number):
             self.assertEqual(x, y.item(), prec, message, allow_inf)
+        elif isinstance(x, torch.BoolTensor) and isinstance(y, torch.BoolTensor):
```
Comment (gchanan) [inline test/common_utils.py]: this looks wrong ... are BoolTensors not instances of Tensors?

File: aten/src/TH/generic/THVector.h [hunk 4]
```diff
@@ -6,6 +6,9 @@
 struct THGenerator;
 
 TH_API void THVector_(fill)(scalar_t *x, const scalar_t c, const ptrdiff_t n);
+
+#if !defined(TH_REAL_IS_BOOL) /* non bool only part */
+
```
Comment (gchanan) [inline aten/src/TH/generic/THVector.h]: can't you end the define at line 24?

File: aten/src/TH/generic/THTensorRandom.cpp [hunk 5]
```diff
@@ -29,6 +29,8 @@ void THTensor_(random)(THTensor *self, THGenerator *_generator)
   TH_TENSOR_APPLY(scalar_t, self, *self_data = (float)(THRandom_random(_generator) % ((1ULL << FLT_MANT_DIG) + 1)););
 #elif defined(TH_REAL_IS_DOUBLE)
   TH_TENSOR_APPLY(scalar_t, self, *self_data = (double)(THRandom_random64(_generator) % ((1ULL << DBL_MANT_DIG) + 1)););
+#elif defined(TH_REAL_IS_BOOL)
+    TH_TENSOR_APPLY(scalar_t, self, *self_data = (bool)(THRandom_random(_generator) % 2););
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorRandom.cpp]: why is this needed?

File: aten/src/TH/generic/THVectorDispatch.cpp [hunk 6]
```diff
@@ -237,4 +239,6 @@ struct THVector_(startup) {
 // Declare a global instance to force static initialization
 static THVector_(startup) THVector_(g_startup);
 
+#endif /* non bool only part */
```
Comment (gchanan) [inline aten/src/TH/generic/THVectorDispatch.cpp]: same comment as the .h.

File: aten/src/TH/generic/THTensorMath.h [hunk 7]
```diff
@@ -5,6 +5,8 @@
 TH_API void THTensor_(fill)(THTensor *r_, scalar_t value);
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorMath.h]: how about moving these functions to THTensorFill or something?

File: torch/csrc/utils/python_numbers.h (hunk 2 above)
Comment (izdeby) [reply torch/csrc/utils/python_numbers.h]: this method was removed in the very last iteration.

File: c10/core/ScalarType.h [hunk 8]
```diff
@@ -40,7 +40,8 @@ _(at::Half,Half,d) \
 _(float,Float,d)   \
 _(double,Double,d) \
 _(std::complex<float>,ComplexFloat,z) \
-_(std::complex<double>,ComplexDouble,z)
+_(std::complex<double>,ComplexDouble,z) \
```
[the remark this answers is not in the document]
Comment (izdeby) [reply c10/core/ScalarType.h]: it is in promote types table as 'b1'
I didn't add bool type to the FORALL_SCALAR_TYPES and SCALAR_TYPES_EXCEPT_HALF as it was not needed for this change. I will add bool to AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF as a part of CUDA change. 

Let me know if you want me to add bool to FORALL_SCALAR_TYPES and SCALAR_TYPES_EXCEPT_HALF and i will do it.

File: aten/src/TH/generic/THVectorDispatch.cpp (hunk 6 above)
Comment (izdeby) [reply aten/src/TH/generic/THVectorDispatch.cpp]: I though about this but i'm not sure if that's the best approach:
1) THVectorDispatch.cpp
It has no header file.
We need THVector_(fill) which requires THVector_(fill_DISPATCHPTR), both are in this file. The problem with moving them out is in THVector_(startup) structure (same file) which initializes all the methods. I agree that having special checks for bool is not good but breaking the whole thing apart is worse as it breaks apart the THVector related methods.

2) THTensorMath.h
Similar story here. We can move the _(fill) thing out but than there is a question what to do with _(zero) that uses it? If we move it, we will end up with a file that has only 2 methods that are somewhat related to each other? Not ideal. My reasoning is that TH will be moved to Aten eventually and it should be cleaned up. 

That's one of the reasons why i rolled back my PR with refactoring stuff to TensorAssignments, it looked messy.

File: aten/src/TH/generic/THTensorRandom.cpp (hunk 5 above)
Comment (izdeby) [reply aten/src/TH/generic/THTensorRandom.cpp]: I'm adding support for the torch.randint which eventually comes down here. If we dont need random, i will revert all random related changes.

File: aten/src/ATen/Dispatch.h [hunk 9]
```diff
@@ -143,6 +143,7 @@
       AT_PRIVATE_CASE_TYPE(at::ScalarType::Long, int64_t, __VA_ARGS__)       \
       AT_PRIVATE_CASE_TYPE(at::ScalarType::Short, int16_t, __VA_ARGS__)      \
       AT_PRIVATE_CASE_TYPE(at::ScalarType::Half, at::Half, __VA_ARGS__)      \
+      AT_PRIVATE_CASE_TYPE(at::ScalarType::Bool, bool, __VA_ARGS__)          \
```
Comment (gchanan) [inline aten/src/ATen/Dispatch.h]: I don't think this solution is scalable; it's totally not obvious what the state of bool is wrt to these macros.  The issues seems to be with Half and Bool.

Can you figure out something better here?

File: torch/csrc/utils/tensor_types.cpp (hunk 1 above)
Comment (izdeby) [reply torch/csrc/utils/tensor_types.cpp]: We need this in order to have torch.BoolTensor
I've updated the code down below to skip creation of the sparse bool type.

File: aten/src/TH/THTensorRandom.cpp [hunk 10]
```diff
@@ -3,3 +3,6 @@
 
 #include <TH/generic/THTensorRandom.cpp>
 #include <TH/THGenerateAllTypes.h>
+
+#include <TH/generic/THTensorRandom.cpp>
```
[the remark this answers is not in the document]
Comment (izdeby) [reply aten/src/TH/THTensorRandom.cpp]: I'm adding support for the torch.randint. If we dont need random, i will revert all random related changes.

File: aten/src/ATen/Declarations.cwrap [hunk 11]
```diff
@@ -2616,6 +2622,7 @@
     - CPU
     - CUDA
   return: self
+  cpu_bool: True
```
[the remark this answers is not in the document]
Comment (izdeby) [reply aten/src/ATen/Declarations.cwrap]: I was under impression that we want to support random() for a bool tensor. If it's not true, i will rollback all random related changes.

File: test/common_utils.py (hunk 3 above)
Comment (izdeby) [reply test/common_utils.py]: You are right, BoolTensor is an instance of Tensor and, lets say, x is 
```
tensor([[False, False],
        [False, False]], dtype=torch.bool)
```

in this case `isinstance(y, torch.Tensor)` will return true, but the rest of compare is written for a integer types. I'm getting an error: "RuntimeError: expected type torch.ByteTensor but got CPUBoolType"  on line 
`diff = a - b`

File: aten/src/TH/generic/THTensorApply.hpp [hunk 12]
```diff
@@ -160,7 +160,7 @@ if (std::isnan(val)) break;
 static inline scalar_t THTensor_(powOne)(scalar_t x, scalar_t y) {
 #if defined(TH_REAL_IS_FLOAT)
   return powf(x, y);
-#elif defined(TH_REAL_IS_DOUBLE)
+#elif defined(TH_REAL_IS_DOUBLE) || defined(TH_REAL_IS_BOOL)
```
[the remark this answers is not in the document]
Comment (izdeby) [reply aten/src/TH/generic/THTensorApply.hpp]: Im getting an error during the CI build: 
/var/lib/jenkins/workspace/aten/src/TH/generic/THTensorApply.hpp: In function 'bool THBoolTensor_powOne(bool, bool)':

/var/lib/jenkins/workspace/aten/src/TH/generic/THTensorApply.hpp:166:16: error: comparison of constant '0' with boolean expression is always true [-Werror=bool-compare]

/var/lib/jenkins/workspace/build/caffe2/aten/src/TH/THGeneral.h:116:35: note: in definition of macro 'THArgCheck'

[the remark this answers is not in the document]
Comment (izdeby) [reply aten/src/TH/generic/THTensorApply.hpp]: As far as i understand, THBoolTensor_powOne is being generated because aten/src/TH/generic/THTensorApply.hpp is being included in the THTensorEvenMoreMath.cpp

Solution to this issue depends on the decision about either _fill & _zero should be moved or not.

File: aten/src/TH/generic/THVectorDispatch.cpp (hunk 6 above)
Comment (izdeby) [reply aten/src/TH/generic/THVectorDispatch.cpp]: In case, if we decide to go ahead and move fill() and zero() from THTensorMath, here is a [PR](https://github.com/pytorch/pytorch/pull/17536)

**Comment by ezyang** [comment]
Test failures are real

File: aten/src/TH/generic/THTensorRandom.cpp (hunk 5 above)
Comment (ezyang) [reply aten/src/TH/generic/THTensorRandom.cpp]: Random for bool SGTM.

File: aten/src/TH/generic/THVectorDispatch.cpp (hunk 6 above)
Comment (ezyang) [reply aten/src/TH/generic/THVectorDispatch.cpp]: I don't think Greg is asking you to move these functions; I think he's referring to "can't you end the define at line 24?" on THVector.h

File: test/common_utils.py [hunk 13]
```diff
@@ -388,6 +388,11 @@ def assertEqual(self, x, y, prec=None, message='', allow_inf=False):
...
+            self.assertEqual(x.storage(), y.storage())
```
Comment (ezyang) [inline test/common_utils.py]: I don't think this is right. If you have a BoolTensor that is a view, storage() will lose that information and give you an incorrect comparison. Better to adjust the code below to take some different strategy with bool is a dtype (and thus we cannot do a subtraction)

File: test/test_torch.py [hunk 14]
```diff
@@ -2750,6 +2764,11 @@ def test_tensor_factory(self):
                 a[0] = 7.
                 self.assertEqual(5., res1[0].item())
 
+        # test boolean tensor
+        a = torch.tensor([True, True, False, True, True], dtype=torch.bool)
+        b = torch.tensor([-1, -1.1, 0, 1, 1.1], dtype=torch.bool)
```
Comment (ezyang) [inline test/test_torch.py]: Not for this PR, but we should probably warn in this case... it seems unlikely that the user actually intended for this to happen

File: test/test_torch.py [hunk 15]
```diff
@@ -2830,6 +2849,19 @@ def test_tensor_factory_cuda_type(self):
         self.assertTrue(x.is_cuda)
         torch.set_default_tensor_type(saved_type)
 
+    # This is a temporary test for a boolean tensors on CPU. Once the CUDA part
+    # will be done, these test cases will be moved down to test_tensor_factories_empty test
+    def test_tensor_factories_empty_bool(self):
+        expectedShape = (1,2)
```
Comment (ezyang) [inline test/test_torch.py]: this fails lint

**Review by ezyang — CHANGES_REQUESTED** [review]
I most heavily reviewed the test code. The main code changes seem basically reasonable.

**Comment by ezyang** [comment]
I took a look at the scalar dispatch macro situation, because I am a bit to blame for the situation.

Here's a summary of what the FORALL macros do:

* `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX` this is the "no actually, this contains every scalar type, and I mean EVERY scalar type" enum. We could probably rename this to `AT_FORALL_SCALAR_TYPES_EVERYTHING` to make its semantics clearer. Note that you very *rarely* use this macro (basically never for kernels), because it contains complex and we don't currently generate kernels for complex ever.

* `AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF` is a *very specialized* macro that was introduced specifically to hack around having to add ComplexHalf accessor support to Scalar. Context is here #11420:

```
    - We grow a even yet more ludicrous macro
      AT_FORALL_SCALAR_TYPES_WITH_COMPLEX_EXCEPT_COMPLEX_HALF
      which does what it says on the tin.  This is because I was
      too lazy to figure out how to define the necessary conversions
      in and out of ComplexHalf without triggering ambiguity problems.
      It doesn't seem to be as simple as just Half.  Leave it for
      when someone actually wants this.
```

* `AT_FORALL_SCALAR_TYPES` is the traditional "everything" macro, predating introduction of complex. It includes half. Since it includes half, it is basically never used to actually do dispatch either, since we don't have half kernels for most things, except in the few cases where we universally support half.

* `AT_FORALL_SCALAR_TYPES_EXCEPT_HALF` is the workhorse macro; it doesn't include complex, it doesn't include half, so if you are working with an operator that doesn't know how to support complex or half specifically, this is the one you'll use.

Hopefully that makes the situation clearer. To summarize, we can roughly classify dtypes into support levels:

* Classic floating point: float, double. Supported universally
* Classic integral: byte, char, int, long, short. Supported universally
* Half. Supported everywhere in CUDA, and nearly nowhere in CPU except a few situations
* Complex. Supported nowhere, except in a few situations
* Bool. I'm actually not sure if the intention is to support universally, or it will be supporting on operators on a case-by-case basis.

File: test/test_torch.py (hunk 14 above)
Comment (izdeby) [reply test/test_torch.py]: Added to my TODO list

File: aten/src/TH/generic/THVectorDispatch.cpp (hunk 6 above)
Comment (izdeby) [reply aten/src/TH/generic/THVectorDispatch.cpp]: oh i see. No, not really. There are a bunch of methods that we dont need for bool but need for all other types, so its ether this way or we have to specify all supported types for _fill() method including bool. Not sure which solution is more clear.

File: aten/src/ATen/Dispatch.h (hunk 9 above)
Comment (izdeby) [reply aten/src/ATen/Dispatch.h]: Could you elaborate, please? I'm not sure if i follow what is wrong here

**Comment by gchanan** [comment]
@ezyang I was actually referring to the dispatch macros here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h

I don't know if we have a name for the macros you are referencing -- ScalarType macros or something?

In any case, I'm less worried about those because users don't usually touch them.  But they do touch the dispatch macros.

**Comment by gchanan** [comment]
how about something like this: consider both Half and Bool special.

So we can keep e.g. AT_DISPATCH_FLOATING_TYPES_AND_HALF because I doubt anyone will ever want AT_DISPATCH_FLOATING_TYPES_AND_HALF_AND_BOOL.  And we can add AT_DISPATCH_INTEGRAL_TYPES_AND_BOOL as the analog of AT_DISPATCH_FLOATING_TYPES_AND_HALF.

The issue is with the AT_DISPATCH_ALL_TYPES* macros.  Instead of having AT_DISPATCH_ALL_TYPES_AND_HALF, how about:
AT_DISPATCH_ALL_TYPES_AND(SCALARTYPE, ...)
AT_DISPATCH_ALL_TYPES_AND(SCALARTYPE1, SCALARTYPE2)

so AT_DISPATCH_ALL_TYPES_AND_HALF would just be AT_DISPATCH_ALL_TYPES_AND(ScalarType::HALF,...) and what would logically have been AT_DISPATCH_ALL_TYPES_AND_HALF_AND_BOOL would be AT_DISPATCH_ALL_TYPES_AND(ScalarType::HALF, ScalarType::BOOL, ...)

you can implement this at statically with templates.

Similarly with AT_DISPATCH_ALL_TYPES_AND_COMPLEX -- just add:
AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(SCALARTYPE,...)
AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND(SCALARTYPE1, SCALARTYPE2,...)

File: test/common_utils.py [hunk 16]
```diff
@@ -392,34 +392,38 @@ def assertEqual(self, x, y, prec=None, message='', allow_inf=False):
...
-                            inf_mask = torch.isinf(a)
-                            inf_sign = inf_mask.sign()
-                            self.assertTrue(torch.equal(inf_sign, torch.isinf(b).sign()), message)
-                            diff[inf_mask] = 0
-                    # TODO: implement abs on CharTensor (int8)
-                    if diff.is_signed() and diff.dtype != torch.int8:
-                        diff = diff.abs()
-                    max_err = diff.max()
-                    self.assertLessEqual(max_err, prec, message)
+                if isinstance(x, torch.BoolTensor) and isinstance(y, torch.BoolTensor):
```
Comment (ezyang) [inline test/common_utils.py]: `isinstance` is the legacy way to test the dtype of a tensor. The modern way is to check if `x.dtype == torch.bool` (or whatever you've called the dtype).

Also, I'd suggest adjusting where exactly you place the conditional here. Notice in the code you short-circuited out, we do a size check first. This is generally useful and you should keep that test even in the boolean case. It seems to that you should put the conditional around `diff = a - b` and subsequent lines.

Comment (ezyang) [reply test/common_utils.py]: Also, defensive programming says that you should explicitly check for the case when only one or the other tensor is bool, and raise an error in that  case (since it means programmer error), rather than let it fall to subtraction and get some sort of weird error that you tried to subtract a float from a bool, e.g.

File: test/test_torch.py [hunk 17]
```diff
@@ -2341,6 +2341,13 @@ def test_zeros(self):
         torch.zeros(100, 100, out=res2)
         self.assertEqual(res1, res2)
 
+        res1 = torch.zeros(2, 2, dtype=torch.bool)
+        expected = torch.tensor([[False, False], [False, False]], dtype=torch.bool)
+        self.assertEqual(res1, expected)
+
+        halfTensor = torch.zeros(1, 1, dtype=torch.half)
+        self.assertEqual(halfTensor, torch.tensor([[0.]], dtype=torch.float16))
```
Comment (ezyang) [inline test/test_torch.py]: The tests here are weirdly asymmetric with the test above. I... guess it's OK? It looks kind of weird though

Comment (ezyang) [reply test/test_torch.py]: (The asymmetry is, the test above is testing that `out=` and non-out implementations match; and you're testing that tensor interprets bool literals in the same way that zeros fills out. I don't mind that much, because the name of the test suggests "general tests for zeros" which, OK, fine, why not.)

**Review by ezyang — APPROVED** [review]
GO GO GO

File: c10/core/ScalarType.h [hunk 18]
```diff
@@ -40,7 +40,8 @@ _(at::Half,Half,d) \
...
+_(bool,Bool,i)
 
 #define AT_FORALL_SCALAR_TYPES(_) \
```
Comment (gchanan) [inline c10/core/ScalarType.h]: I'm not very confident about the lack of Bool here -- can you look at the call sites and see what makes sense to do here?

File: aten/src/TH/generic/THTensorMath.h [hunk 19]
```diff
@@ -91,7 +91,11 @@ TH_API void THTensor_(cmin)(THTensor *r, THTensor *t, THTensor *src);
 TH_API void THTensor_(cmaxValue)(THTensor *r, THTensor *t, scalar_t value);
 TH_API void THTensor_(cminValue)(THTensor *r, THTensor *t, scalar_t value);
 
+TH_API void THTensor_(zerosLike)(THTensor *r_, THTensor *input);
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorMath.h]: are any changes in this file actually needed?  I got rid of all of them and (if I move the implementation of powOne up in the file) it compiles fine on CPU.

File: aten/src/TH/generic/THTensorRandom.cpp [hunk 20]
```diff
@@ -29,6 +29,8 @@ void THTensor_(random)(THTensor *self, THGenerator *_generator)
   TH_TENSOR_APPLY(scalar_t, self, *self_data = (float)(THRandom_random(_generator) % ((1ULL << FLT_MANT_DIG) + 1)););
 #elif defined(TH_REAL_IS_DOUBLE)
   TH_TENSOR_APPLY(scalar_t, self, *self_data = (double)(THRandom_random64(_generator) % ((1ULL << DBL_MANT_DIG) + 1)););
+#elif defined(TH_REAL_IS_BOOL)
```
Comment (gchanan) [inline aten/src/TH/generic/THTensorRandom.cpp]: do we still need geometric in this file?  Your change makes it be generated for bool, but it doesn't seem relevant.

File: aten/src/TH/THTensor.h [hunk 21]
```diff
@@ -21,6 +21,9 @@
 #include <TH/generic/THTensorRandom.h>
 #include <TH/THGenerateAllTypes.h>
 
+#include <TH/generic/THTensorRandom.h>
+#include <TH/THGenerateBoolType.h>
```
Comment (gchanan) [inline aten/src/TH/THTensor.h]: this generates code for random, clampedRandom, and cappedRandom -- do we have tests for all of these with bool?
